managedtenants/bundles/docker_api.py

```diff
@@ -0,0 +1,174 @@
+        debug=False,
+        force_push=False,
+    ):
+        self.quay_api = quay_api if quay_api is not None else QuayAPI()
+        self.registry = (
+            registry if registry is not None else f"quay.io/{self.quay_api.org}"
+        )
+        self.client = docker.from_env()
+        self.dockercfg_path = dockercfg_path
+        self.force_push = force_push
+        self.log = get_text_logger(
+            "managedtenants-docker",
+            level=logging.DEBUG if debug else logging.INFO,
+        )
+
+    def build_bundle(self, path, tag):
+        dockerfile = """
+        FROM scratch
+        COPY manifests /manifests/
+        COPY metadata /metadata/
+        """
+        return self._build(
+            path=path,
+            dockerfile=dockerfile,
+            tag=tag,
+            labels=_parse_metadata_annotations(path),
+        )
+
+    def _build(self, path, dockerfile, tag, labels=None):
+        """
+        Build a docker image using in-memory dockerfile.
+
+        :param path: Path to the context directory.
+        :param dockerfile: A file object to use as the Dockerfile.
+        :param tag: Tag for the built image.
+        :param labels: (Optional) image labels.
+
+        :raise DockerError: on a build error or if the built image has Size 0.
+        """
+        try:
+            out_image, log_generator = self.client.images.build(
+                path=str(path),  # Path(..) obj are not allowed
+                dockerfile=dockerfile,
+                labels=labels if labels is not None else {},
+                tag=tag,
+            )
+            for log in log_generator:
+                self.log.debug(log)
+
+            self.log.debug(out_image.attrs)
+            if out_image.attrs.get("Size", -1) == 0:
+                raise DockerError(f"Built an empty image for path {path}.")
+
+            return out_image
+
+        except docker.errors.BuildError as e:
+            raise DockerError(
+                f"Failed to build image for path {path}, got {e}."
+            )
+
+    def push(self, image):
+        """
+        Push an image to a remote repository.
+
+        :param image: Sretoolbox Image(..) to be pushed.
+
+        :raise DockerError: failed to push image.
+        """
+        try:
+            self._login()
+
+            if self._is_quay_registry():
+                self.quay_api.ensure_repo(image.repository)
+
+            if not self._image_exists(image) or self.force_push:
+                response = self.client.api.push(
+                    image.url_tag, stream=True, decode=True
+                )
+                for log in response:
+                    self.log.debug(log)
+
+        except QuayAPIError as e:
+            raise DockerError(
+                f"Failed to ensure quay repo {image.repository} got {e}."
+            )
+
+        except docker.errors.APIError as e:
+            raise DockerError(f"Failed to push {image.url_tag}, got {e}.")
+
+    def _login(self):
```

Comment:
Did you try without logging in?

Basically login is only required if you don't have a config with credentials already populated.

Otherwise the docker-py client will just add the auth headers to each request from the cred store within the config file it loads.

It does expect the defined env key of `DOCKER_CONFIG` though:

https://github.com/docker/docker-py/blob/a48a5a9647761406d66e8271f19fab7fa0c5f582/docker/utils/config.py#L33-L38

The client doesn't expose a way to define the config location in the constructor so the file is either in an expected location or set by the ENV

Reply:
That worked like a charm :) thank you.